### ISSUE TITLE
Collection Mapping does not work with more than one item

### ISSUE THREAD

**Issue by template-provider** [JsonMapper/JsonMapper#147]
I have something like this:

```php
        $factoryRegistry = FactoryRegistry::withNativePhpClassesAdded();
        $jsonMapper = JsonMapperBuilder::new()
            ->withDocBlockAnnotationsMiddleware()
            ->withObjectConstructorMiddleware($factoryRegistry)
            ->withPropertyMapper(new PropertyMapper($factoryRegistry))
            ->build();

        $jsonMapper->push(new CaseConversion(TextNotation::UNDERSCORE(), TextNotation::CAMEL_CASE()));
```

my classes
```php

<?php

declare(strict_types = 1);

use Delivery;

class DeliveryCollection
{
    /**
     * @param Delivery[] $items
     */
    public function __construct(public readonly array $items = [])
    {
    }
}

<?php

declare(strict_types = 1);

class Delivery
{
    /**
     * @param DeliveryOrder[] $orders
     */
    public function __construct(
        public readonly array $orders = [],
    ) {
    }
}

<?php

declare(strict_types = 1);

class DeliveryOrder
{
    public function __construct(
        public readonly ?int $id,
    ) {
    }
}
```

and a json:
```json
{
    "items": [
        {
            "orders": [
                {
                    "id": 1
                }
            ]
        },
        {
            "orders": [
                {
                    "id": 4
                }
            ]
        }
    ]
}
```

=> TypeError : JsonMapper\Middleware\Constructor\DefaultFactory::__invoke(): Argument #1 ($json) must be of type stdClass, array given, called in DefaultFactory.php:76

If I have only one item it works:
```json
{
    "items": [
        {
            "orders": [
                {
                    "id": 1
                }
            ]
        }
    ]
}
```

**Comment by template-provider** [comment]
ok found the issue .... the bug comes with the issue #142  (another by me)

with this change:
```php
            if ($this->classFactoryRegistry->hasFactory($type)) {
                $value =  $this->classFactoryRegistry->create($type, $value);
            }
```

the factory, that exists in the second item is causing the error

**Comment by template-provider** [comment]
I have a test case for you:

```php
public function testHandleCollectionMapping(): void
    {
        $factoryRegistry = FactoryRegistry::withNativePhpClassesAdded();
        $jsonMapper = JsonMapperBuilder::new()
            ->withDocBlockAnnotationsMiddleware()
            ->withObjectConstructorMiddleware($factoryRegistry)
            ->withPropertyMapper(new PropertyMapper($factoryRegistry))
            ->build();

        $jsonMapper->push(new CaseConversion(TextNotation::UNDERSCORE(), TextNotation::CAMEL_CASE()));
        
        $json = '{
    "items": [
        {
            "name": "foo",
            "orders": [
                {
                    "name": "bar"
                }
            ]
        },
        {
            "name": "foo",
            "orders": [
                {
                    "name": "bar"
                }
            ]
        }
    ]
}';

        $result = $jsonMapper->mapToClass((object) json_decode($json), FooCollection::class);
        
        static::assertInstanceOf(FooCollection::class, $result);
    }
```

```php
<?php

declare(strict_types = 1);

namespace Tests\Component\Foo;

class FooCollection
{
    /**
     * @param FooItem[] $items
     */
    public function __construct(public readonly array $items = [])
    {
    }
}
```

```php
<?php

declare(strict_types = 1);

namespace Tests\Component\Foo;

class FooItem
{
    /**
     * @param BarItem[] $orders
     */
    public function __construct(public readonly string $name, public readonly array $orders = [],)
    {
    }
}
```

```php
<?php

declare(strict_types = 1);

namespace Tests\Component\Foo;

class BarItem
{
    public function __construct(
        public readonly ?int $id,
    ) {
    }
}

```

TypeError : JsonMapper\Middleware\Constructor\DefaultFactory::__invoke(): Argument #1 ($json) must be of type stdClass, array given, called in

=> 2nd item in test case is causing this error

**Comment by template-provider** [comment]
it seems with a constructor parameter of the 2nd item this issues happens

**Comment by template-provider** [comment]
any updates? thx

**Comment by DannyvdSluijs** [comment]
I'll try to see if I can find a fix. To be honest JsonMapper is a spare time project and with a busy work and family live at the moment working on JsonMapper tends to disappear to the background. 

This doesn't change the fact I'm happy to see actively involved users providing me with test cases pointing out their issues. So you effect is valued and not in vain. It just might take a bit longer than you're used to from the past issues.

**Comment by DannyvdSluijs** [comment]
@template-provider I've been working on this putting effort in aligning the way values are being created in the `DefaultFactory`. This refactor will require some more work on the testing which I'm more than happy to do as this was already something I wanted to fix anyway.

I was hoping you could help out a bit though. Could you test with the branch containing the changes? You should be able to install it using `composer require json-mapper/json-mapper:dev-Bug-147`. This should fix your problem, as your provided testcases are in, but I would like to know if everything works for you.

**Comment by template-provider** [comment]
@DannyvdSluijs 
can do  .... give me until next week and I will test it

**Comment by template-provider** [comment]
@DannyvdSluijs 
still have the problems ... even with the bug branch and the original test:

Testing started at 08:17 ...
PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.1.13
Configuration: phpunit.xml.dist


JsonMapper\Exception\TypeError : JsonMapper\JsonMapper::mapToClass(): Argument #2 ($class) must be of type class-string, string given, called in

**Comment by DannyvdSluijs** [comment]
@template-provider the error message has changed so maybe something else broke in the refactor. Could you share the full stack trace with me? Also your test is part of the PR, see https://github.com/JsonMapper/JsonMapper/pull/153/files#diff-6717905da265af10bcc76381f7c97541e5421900a56fd0325df59fe49f9211dbR186 and it is passing all builds.

Maybe we are onto another issue?

**Comment by template-provider** [comment]
@DannyvdSluijs 
Ok ... tested it with both repos and found the problem. the branch works ... it has some problems with the autoload+namespace lookup => results in error.
my namespace was not 100% compat with psr-0/4 for the test in my client.
but in general with the classmap autoload it should work .... little bit strange.

but in general I would say: works as designed ... even with my "$jsonMapper->push(new CaseConversion(TextNotation::UNDERSCORE(), TextNotation::CAMEL_CASE()));" changes.

I think we can merge and close this

**Comment by template-provider** [comment]
@DannyvdSluijs can you plz create a tag ... would be awesome

**Comment by DannyvdSluijs** [comment]
Hi @template-provider

I'll continue to increase to coverage in the next couple of days. Once I've reached the same percentage I can merge it and create a new release straight after it. Should be soon.

**Comment by DannyvdSluijs** [comment]
Just released 2.15.0
https://github.com/JsonMapper/JsonMapper/releases/tag/2.15.0
